### ISSUE TITLE
SIGSEGV: Crash on local replication

### ISSUE THREAD

**Issue by mizhka** [zrepl/zrepl#173]
Hi,

My first zrepl daemon crashed after few hours of replication. 

```
Apr 22 23:00:17 gidrarium zrepl[1279]: job=local_sink subsystem=transport msg="error responding to client" err="client-provided callback did block on send" 
Apr 22 23:00:17 gidrarium zrepl[1279]: job=local_sink subsystem=transportmux msg="accept error" err="client-provided callback did block on send" errType=*errors.errorString 
Apr 22 23:00:17 gidrarium zrepl[1278]: 2019-04-22T23:00:17+03:00 [ERRO][local_sink]: error responding to client subsystem="transport" err="client-provided callback did block on send"
Apr 22 23:00:17 gidrarium zrepl[1279]: job=backup_system subsystem=rpc.data msg="ReqRecv returns" invocation=106 
Apr 22 23:00:17 gidrarium zrepl[1278]: 2019-04-22T23:00:17+03:00 [ERRO][local_sink]: accept error err="client-provided callback did block on send" subsystem="transportmux" errType="*errors.errorString"
Apr 22 23:00:17 gidrarium zrepl[1278]: panic: runtime error: invalid memory address or nil pointer dereference
Apr 22 23:00:17 gidrarium zrepl[1278]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x905554]
Apr 22 23:00:17 gidrarium zrepl[1278]: 
Apr 22 23:00:17 gidrarium zrepl[1278]: goroutine 18267 [running]:
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/rpc/versionhandshake.DoHandshakeVersion(0x0, 0x0, 0xbf27a536f1189fad, 0x1cacc8f8138d, 0x10a20c0, 0x1, 0x0)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/rpc/versionhandshake/versionhandshake.go:170 +0x134
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/rpc/versionhandshake.DoHandshakeCurrentVersion(0x0, 0x0, 0xbf27a536f1189fad, 0x1cacc8f8138d, 0x10a20c0, 0x1cacc8f8138d)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/rpc/versionhandshake/versionhandshake.go:155 +0x5c
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/rpc/versionhandshake.HandshakeConnecter.Connect(0xbf8700, 0xc000168880, 0x2540be400, 0xc00900, 0xc000524440, 0xc0001b38c0, 0xa72600, 0x458bd0, 0xc0000d9808)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/rpc/versionhandshake/versionhandshake_transport_wrappers.go:25 +0x119
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/rpc/transportmux.labeledConnecter.Connect(0xc00016a040, 0x40, 0x40, 0xbf9ae0, 0xc000522020, 0xc00900, 0xc000524440, 0x0, 0x0, 0x0, ...)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/rpc/transportmux/transportmux.go:164 +0x76
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/rpc/dataconn.(*Client).getWire(0xc000522140, 0xc00900, 0xc000524440, 0xb4e74f, 0xf, 0x20)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/rpc/dataconn/dataconn_client.go:98 +0x4c
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/rpc/dataconn.(*Client).ReqRecv(0xc000522140, 0xc00900, 0xc000524440, 0xc000368a00, 0x803ea4328, 0xc0000aefa0, 0x0, 0x0, 0x0)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/rpc/dataconn/dataconn_client.go:145 +0xbc
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/rpc.(*Client).Receive(0xc000130120, 0xc00900, 0xc000524440, 0xc000368a00, 0x803ea4328, 0xc0000aefa0, 0xc0000aef60, 0x0, 0x0)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/rpc/rpc_client.go:100 +0x67
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/replication/logic.(*Step).doReplication(0xc0002e3540, 0xc00900, 0xc000524440, 0x0, 0x0)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/replication/logic/replication_logic.go:470 +0x35b
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/replication/logic.(*Step).Step(0xc0002e3540, 0xc00900, 0xc000524440, 0x0, 0xed44f9cba)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/replication/logic/replication_logic.go:171 +0x3f
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/replication/driver.(*fs).do.func2()
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/replication/driver/replication_driver.go:453 +0x103
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/util/chainlock.(*L).DropWhile(0xc0001ba770, 0xc0000d9f28)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/util/chainlock/chainlock.go:41 +0x5b
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/replication/driver.(*fs).do(0xc0000cb270, 0xc00900, 0xc000524440, 0xc000095270, 0x0)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/replication/driver/replication_driver.go:450 +0x82f
Apr 22 23:00:17 gidrarium zrepl[1278]: github.com/zrepl/zrepl/replication/driver.(*attempt).do.func2(0xc0002ee370, 0xc00900, 0xc000524440, 0xc000095270, 0xc0001b58f0, 0xc0000cb270)
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/replication/driver/replication_driver.go:364 +0xa4
Apr 22 23:00:17 gidrarium zrepl[1278]: created by github.com/zrepl/zrepl/replication/driver.(*attempt).do
Apr 22 23:00:17 gidrarium zrepl[1278]:  /repo/freebsd/sandbox-ports/sysutils/zrepl/work/src/github.com/zrepl/zrepl/replication/driver/replication_driver.go:362 +0xf57

```

My config is based on local replication sample:

```
jobs:
  - type: sink
    name: "local_sink"
    root_fs: "miniusb/zrepl/sink/test"
    serve:
      type: local
      listener_name: localsink

  - type: push
    name: "backup_system"
    connect:
      type: local
      listener_name: localsink
      client_identity: local_backup
    filesystems: {
      "tank/bootenv/10/var/log<": true,
    }
    snapshotting:
      type: periodic
      interval: 5m
      prefix: zrepl_
    pruning:
      keep_sender:
      - type: not_replicated
      - type: last_n
        count: 10
      keep_receiver:
      - type: grid
        grid: 1x1h(keep=all) | 24x1h | 35x1d | 6x30d
        regex: "zrepl_.*"
(END)

```

What I've done: export zpool miniusb , wait for few hours, import it back, send signal to wake up job. In few seconds it had been crashed.

**Comment by problame** [comment]
Sry for the late reply...

> What I've done: export zpool miniusb , wait for few hours, import it back, send signal to wake up job. In few seconds it had been crashed.

Could you reproduce this problem following this procedure?

Does the issue persist if you use a [local TCP listener](https://zrepl.github.io/configuration/transports.html#tcp-transport) instead?

It seems to me like 
https://github.com/zrepl/zrepl/blob/afed7627749bee3726a06f7855a11a82b6769b09/transport/local/serve_local.go#L132
would need to return `left == nil` to cause this crash with the `local` transport.

**Comment by problame** [comment]
Commit 0e419c1 contains a fix for the local transport.
Would you mind testing it? You can use the binary artifacts produced by the CI (click on the checkmark next to the commit above) and just replace the one installed by the FreeBSD package.

**Comment by monwarez** [comment]
Thanks, I just upgrade my zrepl version with commit 0e419c1 and currently testing with local transport. 
So far it didn't crash (4 hours run), I will see tomorrow if it crash or not during the night, and keep you updated.

**Comment by problame** [comment]
OK, fixed a comment in 4b0ff80 and will merge into master then.
